File: Comprehensive Health Data Analytics for Chronic Disease Prediction/python_scripts/univariate_bivariate_multivariate.py
"""
Univariate, Bivariate, and Multivariate Analysis
Comprehensive analysis of health data
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

# Set style
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (14, 8)

def load_and_clean_data():
    """Load and clean the health data"""
    df = pd.read_csv('../data/health_data.csv')
    if 'Unnamed: 0' in df.columns:
        df = df.drop('Unnamed: 0', axis=1)
    df['age_years'] = df['age'] / 365.25
    df['bmi'] = df['weight'] / ((df['height'] / 100) ** 2)
    df = df[(df['ap_hi'] >= 80) & (df['ap_hi'] <= 250)]
    df = df[(df['ap_lo'] >= 40) & (df['ap_lo'] <= 150)]
    df = df[df['ap_hi'] >= df['ap_lo']]
    df = df[(df['height'] >= 100) & (df['height'] <= 220)]
    df = df[(df['weight'] >= 30) & (df['weight'] <= 200)]
    df = df[(df['bmi'] >= 10) & (df['bmi'] <= 60)]
    return df

def univariate_analysis(df):
    """Perform univariate analysis"""
    print("=" * 80)
    print("UNIVARIATE ANALYSIS")
    print("=" * 80)
    
    # Numerical variables
    numerical_cols = ['age_years', 'height', 'weight', 'ap_hi', 'ap_lo', 'bmi']
    fig, axes = plt.subplots(2, 3, figsize=(18, 12))
    axes = axes.ravel()
    
    for i, col in enumerate(numerical_cols):
        axes[i].hist(df[col], bins=50, density=True, alpha=0.7, edgecolor='black')
        df[col].plot.density(ax=axes[i], color='red', linewidth=2)
        axes[i].set_title(f'{col.upper()} Distribution', fontsize=12, fontweight='bold')
        axes[i].set_xlabel(col)
        axes[i].set_ylabel('Density')
        axes[i].grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig('../figures/univariate_numerical.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Categorical variables
    categorical_cols = ['gender', 'cholesterol', 'gluc', 'smoke', 'alco', 'active', 'cardio']
    fig, axes = plt.subplots(3, 3, figsize=(18, 15))
    axes = axes.ravel()
    
    for i, col in enumerate(categorical_cols):
        value_counts = df[col].value_counts().sort_index()
        axes[i].bar(value_counts.index.astype(str), value_counts.values, color='steelblue', edgecolor='black')
        axes[i].set_title(f'{col.upper()} Distribution', fontsize=12, fontweight='bold')
        axes[i].set_xlabel(col)
        axes[i].set_ylabel('Count')
        axes[i].grid(True, alpha=0.3, axis='y')
    
    plt.tight_layout()
    plt.savefig('../figures/univariate_categorical.png', dpi=300, bbox_inches='tight')
    plt.close()
    print("Univariate analysis completed. Figures saved.")

def bivariate_analysis(df):
    """Perform bivariate analysis"""
    print("=" * 80)
    print("BIVARIATE ANALYSIS")
    print("=" * 80)
    
    # Scatter plots
    fig, axes = plt.subplots(2, 2, figsize=(16, 12))
    axes = axes.ravel()
    
    pairs = [('age_years', 'bmi'), ('ap_hi', 'ap_lo'), ('weight', 'height'), ('ap_hi', 'bmi')]
    for i, (x, y) in enumerate(pairs):
        axes[i].scatter(df[x], df[y], alpha=0.3, s=10)
        axes[i].set_xlabel(x, fontsize=11)
        axes[i].set_ylabel(y, fontsize=11)
        axes[i].set_title(f'{x.upper()} vs {y.upper()}', fontsize=12, fontweight='bold')
        axes[i].grid(True, alpha=0.3)
        
        corr = df[[x, y]].corr().iloc[0, 1]
        axes[i].text(0.05, 0.95, f'r = {corr:.3f}', transform=axes[i].transAxes, 
                    fontsize=11, verticalalignment='top', 
                    bbox=dict(boxstyle='round', facecolor='wheat', alpha=0.5))
    
    plt.tight_layout()
    plt.savefig('../figures/bivariate_scatter.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Box plots
    fig, axes = plt.subplots(2, 3, figsize=(18, 12))
    axes = axes.ravel()
    
    comparisons = [('bmi', 'cardio'), ('age_years', 'cardio'), ('ap_hi', 'cardio'),
                   ('bmi', 'gender'), ('ap_hi', 'cholesterol'), ('bmi', 'smoke')]
    
    for i, (numerical, categorical) in enumerate(comparisons):
        sns.boxplot(data=df, x=categorical, y=numerical, ax=axes[i])
        axes[i].set_title(f'{numerical.upper()} by {categorical.upper()}', fontsize=12, fontweight='bold')
        axes[i].set_xlabel(categorical)
        axes[i].set_ylabel(numerical)
        axes[i].grid(True, alpha=0.3, axis='y')
    
    plt.tight_layout()
    plt.savefig('../figures/bivariate_boxplots.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Heatmaps
    fig, axes = plt.subplots(2, 3, figsize=(18, 12))
    axes = axes.ravel()
    
    categorical_vars = ['gender', 'cholesterol', 'gluc', 'smoke', 'alco', 'active']
    for i, var in enumerate(categorical_vars):
        crosstab = pd.crosstab(df[var], df['cardio'], normalize='index') * 100
        sns.heatmap(crosstab, annot=True, fmt='.1f', cmap='YlOrRd', ax=axes[i], cbar_kws={'label': '%'})
        axes[i].set_title(f'{var.upper()} vs Cardio', fontsize=12, fontweight='bold')
        axes[i].set_xlabel('Cardio')
        axes[i].set_ylabel(var)
    
    plt.tight_layout()
    plt.savefig('../figures/bivariate_heatmaps.png', dpi=300, bbox_inches='tight')
    plt.close()
    print("Bivariate analysis completed. Figures saved.")

def multivariate_analysis(df):
    """Perform multivariate analysis"""
    print("=" * 80)
    print("MULTIVARIATE ANALYSIS")
    print("=" * 80)
    
    # Correlation heatmap
    numerical_cols = ['age_years', 'height', 'weight', 'ap_hi', 'ap_lo', 'bmi', 'cardio']
    correlation_matrix = df[numerical_cols].corr()
    
    plt.figure(figsize=(12, 10))
    sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', center=0, 
                square=True, linewidths=1, cbar_kws={'shrink': 0.8}, fmt='.3f')
    plt.title('Multivariate Correlation Matrix', fontsize=14, fontweight='bold')
    plt.tight_layout()
    plt.savefig('../figures/multivariate_correlation.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Pair plot
    key_vars = ['age_years', 'bmi', 'ap_hi', 'ap_lo', 'cardio']
    pair_df = df[key_vars].sample(n=min(5000, len(df)), random_state=42)
    
    sns.pairplot(pair_df, hue='cardio', diag_kind='kde', palette=['skyblue', 'salmon'])
    plt.suptitle('Pair Plot: Key Variables by Cardio Status', y=1.02, fontsize=14, fontweight='bold')
    plt.tight_layout()
    plt.savefig('../figures/multivariate_pairplot.png', dpi=300, bbox_inches='tight')
    plt.close()
    print("Multivariate analysis completed. Figures saved.")

if __name__ == "__main__":
    df = load_and_clean_data()
    univariate_analysis(df)
    bivariate_analysis(df)
    multivariate_analysis(df)
    print("\nAll analyses completed successfully!")

File: Comprehensive Health Data Analytics for Chronic Disease Prediction/r_scripts/univariate_bivariate_multivariate.R
# Univariate, Bivariate, and Multivariate Analysis
# Comprehensive analysis of health data

# Load libraries
library(ggplot2)
library(dplyr)
library(corrplot)
library(GGally)

# Load data
df <- read.csv("../data/health_data.csv", stringsAsFactors = FALSE)

# Data cleaning
if ("Unnamed..0" %in% colnames(df)) {
  df <- df[, !colnames(df) %in% "Unnamed..0"]
}

df$age_years <- df$age / 365.25
df$bmi <- df$weight / ((df$height / 100) ^ 2)
df <- df[df$ap_hi >= 80 & df$ap_hi <= 250, ]
df <- df[df$ap_lo >= 40 & df$ap_lo <= 150, ]
df <- df[df$ap_hi >= df$ap_lo, ]
df <- df[df$height >= 100 & df$height <= 220, ]
df <- df[df$weight >= 30 & df$weight <= 200, ]
df <- df[df$bmi >= 10 & df$bmi <= 60, ]

# ============================================================================
# 1. UNIVARIATE ANALYSIS
# ============================================================================

cat("=", rep("=", 78), "\n", sep="")
cat("UNIVARIATE ANALYSIS\n")
cat("=", rep("=", 78), "\n", sep="")

numerical_cols <- c("age_years", "height", "weight", "ap_hi", "ap_lo", "bmi")

# Histograms and density plots
png("../figures/univariate_numerical.png", width = 1800, height = 1200, res = 300)
par(mfrow = c(2, 3))
for (col in numerical_cols) {
  hist(df[[col]], breaks = 50, probability = TRUE, 
       main = paste(toupper(col), "Distribution"),
       xlab = col, col = "steelblue", border = "black")
  lines(density(df[[col]]), col = "red", lwd = 2)
}
dev.off()

# Categorical variables
categorical_cols <- c("gender", "cholesterol", "gluc", "smoke", "alco", "active", "cardio")

png("../figures/univariate_categorical.png", width = 1800, height = 1500, res = 300)
par(mfrow = c(3, 3))
for (col in categorical_cols) {
  value_counts <- table(df[[col]])
  barplot(value_counts, main = paste(toupper(col), "Distribution"),
          xlab = col, ylab = "Count", col = "steelblue", border = "black")
}
dev.off()

# ============================================================================
# 2. BIVARIATE ANALYSIS
# ============================================================================

cat("=", rep("=", 78), "\n", sep="")
cat("BIVARIATE ANALYSIS\n")
cat("=", rep("=", 78), "\n", sep="")

# Scatter plots
pairs_list <- list(
  c("age_years", "bmi"),
  c("ap_hi", "ap_lo"),
  c("weight", "height"),
  c("ap_hi", "bmi")
)

png("../figures/bivariate_scatter.png", width = 1600, height = 1200, res = 300)
par(mfrow = c(2, 2))
for (pair in pairs_list) {
  x <- pair[1]
  y <- pair[2]
  plot(df[[x]], df[[y]], pch = 19, cex = 0.3, alpha = 0.5,
       main = paste(toupper(x), "vs", toupper(y)),
       xlab = x, ylab = y)
  corr_val <- cor(df[[x]], df[[y]], use = "complete.obs")
  legend("topright", legend = paste("r =", round(corr_val, 3)), 
         bty = "n", bg = "wheat")
  grid()
}
dev.off()

# Box plots
comparisons <- list(
  c("bmi", "cardio"),
  c("age_years", "cardio"),
  c("ap_hi", "cardio"),
  c("bmi", "gender"),
  c("ap_hi", "cholesterol"),
  c("bmi", "smoke")
)

png("../figures/bivariate_boxplots.png", width = 1800, height = 1200, res = 300)
par(mfrow = c(2, 3))
for (comp in comparisons) {
  numerical <- comp[1]
  categorical <- comp[2]
  boxplot(df[[numerical]] ~ df[[categorical]],
          main = paste(toupper(numerical), "by", toupper(categorical)),
          xlab = categorical, ylab = numerical,
          col = "steelblue", border = "black")
  grid(nx = NA, ny = NULL)
}
dev.off()

# Heatmaps
categorical_vars <- c("gender", "cholesterol", "gluc", "smoke", "alco", "active")

png("../figures/bivariate_heatmaps.png", width = 1800, height = 1200, res = 300)
par(mfrow = c(2, 3))
for (var in categorical_vars) {
  crosstab <- prop.table(table(df[[var]], df$cardio), margin = 1) * 100
  heatmap(crosstab, main = paste(toupper(var), "vs Cardio"),
          xlab = "Cardio", ylab = var, col = heat.colors(256),
          Rowv = NA, Colv = NA)
}
dev.off()

# ============================================================================
# 3. MULTIVARIATE ANALYSIS
# ============================================================================

cat("=", rep("=", 78), "\n", sep="")
cat("MULTIVARIATE ANALYSIS\n")
cat("=", rep("=", 78), "\n", sep="")

# Correlation matrix
numerical_cols <- c("age_years", "height", "weight", "ap_hi", "ap_lo", "bmi", "cardio")
correlation_matrix <- cor(df[numerical_cols], use = "complete.obs")

png("../figures/multivariate_correlation.png", width = 1200, height = 1000, res = 300)
corrplot(correlation_matrix, method = "color", type = "upper",
         order = "hclust", tl.cex = 0.8, tl.col = "black",
         addCoef.col = "black", number.cex = 0.7,
         title = "Multivariate Correlation Matrix")
dev.off()

# Pair plot
key_vars <- c("age_years", "bmi", "ap_hi", "ap_lo", "cardio")
pair_df <- df[sample(nrow(df), min(5000, nrow(df))), key_vars]
pair_df$cardio <- as.factor(pair_df$cardio)

png("../figures/multivariate_pairplot.png", width = 2000, height = 2000, res = 300)
ggpairs(pair_df, aes(color = cardio, alpha = 0.7),
        lower = list(continuous = "points"),
        upper = list(continuous = "cor"),
        diag = list(continuous = "densityDiag"))
dev.off()

cat("\nUnivariate, bivariate, and multivariate analysis completed successfully!\n")

